moveit_ros/planning_interface/move_group_interface/src/move_group_interface.cpp

```diff
@@ -345,7 +345,7 @@ class MoveGroupInterface::MoveGroupInterfaceImpl
     max_acceleration_scaling_factor_ = max_acceleration_scaling_factor;
   }
 
-  robot_state::RobotState& getTargetRobotState()
+  robot_state::RobotState& getTargetRobotState() const
```

Comment:
This looks like a mistake to me? Technically it does not modify the MGI, but it returns a non-const reference to an important internally used `RobotState` object which is dynamically allocated.

Reply:
I admit I do not understand the intent behind this function. But maybe there is a use case where giving write access to the RobotState is justified. As for the constness of this function, it does not modify the logical state and therefore it can and should be marked as const. What happens to the internal state of the RobotState object is decided at call site.

A possible fix to this would be to return a const reference but that might be a breaking change for those code bases (whatever they may be) that rely on obtaining write access on robot state through this function.

Reply:
> it does not modify the logical state and therefore it can and should be marked as const. What happens to the internal state of the RobotState object is decided at call site, not in this function.

This argument is plain wrong.

```
struct A {
   int x = 42;
   int& f(){ return x;};
};

int main(){
   const A a;
   return a.f();
}
```

does not compile, the only reason why it compiles here is that the compiler does not make assumptions about pointed-to objects.
The shared pointer of this object is exclusively owned by the MGI (it should actually be a `unique_ptr` if you want to state this explicitly), so it's part of the state of it.

Reply:
I don't quite follow your reasoning here. Consider the following code snippet which better reflects the situation at hand here:
```
#include <memory>

struct A {
   std::shared_ptr<int> x = std::make_shared<int>(42);
   int& f() const { return *x;};
};

int main(){
   const A a;
   return a.f();
}
```
Now, if A is indeed marked as const, the code will not compile without the const qualifier in function f. If it is marked as non-const then the constness of function f does not make any difference from compiling point of view. In MoveIt codebase we are simply returning a reference to a dereferenced shared_ptr. Sure you can modify the object that you receive but that is up to you a caller. If the idea is to prevent that, then only read access must be provided by returning a const reference or copy of the state by returning by value.

Reply:
> Sure you can modify the object that you receive but that is up to you as a caller

The object uniquely belongs to the parent object, be it a member variable or a pointer to an object, where the pointer cannot be accessed through interfaces.
In this case returning a non-const reference to the member discards const of the parent object.

The correct way to provide the function for read and r&w access is a const overload, as @davetcoleman stated too. Although the `NonConst` suffix is neither needed nor nice to have:
```
struct A {
   unique_ptr<int> x;
   A() :  x(make_unique<int>(42)) {}
   int& f() { return *x; }
   const int& f() const { return *x; }
};
```

We currently do not define the const overload for `getTargetRobotState()`, because nobody ever tried to use this function on a const MGI (or even use a const MGI).
The current specification, however, is clearly non-const in a way that is not automatically inferred by the compiler (as it does not track ownership of pointed-to objects).

Please remove the const here, and, if you feel a const version is needed, provide an additional `const` overload.